fft and ifft implementation

joglekara opened the pull request:
Please see https://github.com/google/jax/issues/1877 for reference

Commit:
first commit with placeholders for tests

joglekara commented:
Happy to have some feedback to know if I'm on the right track. 

I'm going to flesh out the tests next

shoyer reviewed (commented):
This looks like a good start. I suspect your implementations are correct, but we'll need tests to be sure (which is probably where most of the work will be).

Commit:
added tests for the following:
1 - inverse
2 - dtypes
3 - size
4 - axis

added error tests for the following:
1 - multiple axes provided instead of single axis
2 - axis out of bounds

joglekara commented:
Looks like the tests made it through the CI build. 

I should add the documentation with this same PR, right?

shoyer commented on tests/fft_test.py:
```diff
@@ -105,6 +105,64 @@ def testFftnErrors(self, inverse):
+        tol = 0.15  # TODO(skye): can we be more precise?
+        jtu.check_grads(np_fn, args_maker(), order=1, atol=tol, rtol=tol)
+        jtu.check_grads(np_fn, args_maker(), order=2, atol=tol, rtol=tol)
+    pass
```
delete?

joglekara replied:
good catch

shoyer commented on tests/fft_test.py:
```diff
@@ -105,6 +105,64 @@ def testFftnErrors(self, inverse):
+      for size in [10]
+      for axis in [-1, 0]))
+  def testFft(self, inverse, size, dtype, axis, rng_factory):
+    # placeholder for fft/ifft tests
```
delete?

joglekara replied:
good catch

shoyer commented on jax/numpy/fft.py:
```diff
@@ -92,6 +92,36 @@ def ifftn(a, s=None, axes=None, norm=None):
+      "Please use jax.np.fft.fftn. "
+      "Got axis %s." % (list(axis))
+    )
+
```
This is probably fine, but maybe unnecessary, as long as a ValueError or TypeError gets raised instead of computing the wrong thing.

skye replied:
Also as long as there's a clear error message! `testFftErrors` makes this easy to check, thanks for adding it.

joglekara replied:
Sounds like a keeper then, thanks all.

shoyer reviewed (commented):
Yes, please add these to the list in https://github.com/google/jax/blob/master/docs/jax.numpy.rst to ensure they appear in the online docs!

Commit:
removed placeholders
added functions to .rst file

joglekara commented:
pushed the requested changes

shoyer commented:
Thanks Archis!
